File: core/game.test.ts
import { assert } from "@std/assert"
import { expect } from "@std/expect"
import { describe, it } from "@std/testing/bdd"
import { Either } from "effect"

import { Game, TooFewPlayersError, TooManyPlayersError } from "./game.ts"

describe("#initialize", () => {
	it("should return an error if there are too few players", () => {
		const result = Game.new(["Alice"])

		expect(result).toEqual(Either.left(new TooFewPlayersError(1)))
	})

	it("should give five cards to each player if there are four or more players", () => {
		const gameResult = Game.new(["Alice", "Bob", "Charlie", "David"])

		assert(
			Either.isRight(gameResult),
			"Expected Either.right, got Either.left",
		)

		const { playersStillAlive } = gameResult.right

		expect(playersStillAlive).toHaveLength(4)

		playersStillAlive.forEach((player) => {
			expect(player.tryalCardsRemaining).toEqual(5)
		})
	})

	it("should give five cards to each player if there are six or less players", () => {
		const gameResult = Game.new([
			"Alice",
			"Bob",
			"Charlie",
			"David",
			"Eve",
			"Frank",
		])

		assert(
			Either.isRight(gameResult),
			"Expected Either.right, got Either.left",
		)

		const { playersStillAlive } = gameResult.right

		expect(playersStillAlive).toHaveLength(6)

		playersStillAlive.forEach((player) => {
			expect(player.tryalCardsRemaining).toEqual(5)
		})
	})

	it("should give four cards to each player if there are seven or more players", () => {
		const gameResult = Game.new([
			"Alice",
			"Bob",
			"Charlie",
			"David",
			"Eve",
			"Frank",
			"Grace",
		])

		assert(
			Either.isRight(gameResult),
			"Expected Either.right, got Either.left",
		)

		const { playersStillAlive } = gameResult.right

		expect(playersStillAlive).toHaveLength(7)

		playersStillAlive.forEach((player) => {
			expect(player.tryalCardsRemaining).toEqual(4)
		})
	})

	it("should give four cards to each player if there are nine or less players", () => {
		const gameResult = Game.new([
			"Alice",
			"Bob",
			"Charlie",
			"David",
			"Eve",
			"Frank",
			"Grace",
			"Heidi",
			"Ivan",
		])

		assert(
			Either.isRight(gameResult),
			"Expected Either.right, got Either.left",
		)

		const { playersStillAlive } = gameResult.right

		expect(playersStillAlive).toHaveLength(9)

		playersStillAlive.forEach((player) => {
			expect(player.tryalCardsRemaining).toEqual(4)
		})
	})

	it("should give three cards to each player if there are ten or more players", () => {
		const gameResult = Game.new([
			"Alice",
			"Bob",
			"Charlie",
			"David",
			"Eve",
			"Frank",
			"Grace",
			"Heidi",
			"Ivan",
			"Judy",
		])

		assert(
			Either.isRight(gameResult),
			"Expected Either.right, got Either.left",
		)

		const { playersStillAlive } = gameResult.right

		expect(playersStillAlive).toHaveLength(10)

		playersStillAlive.forEach((player) => {
			expect(player.tryalCardsRemaining).toEqual(3)
		})
	})

	it("should give three cards to each player if there are twelve or less players", () => {
		const gameResult = Game.new([
			"Alice",
			"Bob",
			"Charlie",
			"David",
			"Eve",
			"Frank",
			"Grace",
			"Heidi",
			"Ivan",
			"Judy",
			"Kevin",
			"Laura",
		])

		assert(
			Either.isRight(gameResult),
			"Expected Either.right, got Either.left",
		)

		const { playersStillAlive } = gameResult.right

		expect(playersStillAlive).toHaveLength(12)

		playersStillAlive.forEach((player) => {
			expect(player.tryalCardsRemaining).toEqual(3)
		})
	})

	it("should return an error if there are too many players", () => {
		const result = Game.new([
			"Alice",
			"Bob",
			"Charlie",
			"David",
			"Eve",
			"Frank",
			"Grace",
			"Heidi",
			"Ivan",
			"Judy",
			"Kevin",
			"Laura",
			"Mallory",
		])

		expect(result).toEqual(Either.left(new TooManyPlayersError(13)))
	})
})

describe("#revealTryalCard", () => {
	it("should reduce the card amount of the targeted player by one if the revealed card is not a witch", () => {
		const gameResult = Game.new(["Alice", "Bob", "Charlie", "David"])

		assert(
			Either.isRight(gameResult),
			"Expected Either.right, got Either.left",
		)

		const { right: game } = gameResult

		const updatedGame = game.revealTryalCard(1, false)

		const { playersStillAlive } = updatedGame

		expect(playersStillAlive).toHaveLength(4)

		playersStillAlive.forEach((player, position) => {
			if (position === 1) {
				expect(player.name).toEqual("Bob")
				expect(player.tryalCardsRemaining).toEqual(4)
			} else {
				expect(player.name).not.toEqual("Bob")
				expect(player.tryalCardsRemaining).toEqual(5)
			}
		})
	})

	it("should kill the targeted player if the revealed card is a witch", () => {
		const gameResult = Game.new(["Alice", "Bob", "Charlie", "David"])

		assert(
			Either.isRight(gameResult),
			"Expected Either.right, got Either.left",
		)

		const { right: game } = gameResult

		const updatedGame = game.revealTryalCard(1, true)

		const { playersStillAlive } = updatedGame

		expect(playersStillAlive).toHaveLength(3)

		playersStillAlive.forEach((player) => {
			expect(player.name).not.toEqual("Bob")
			expect(player.tryalCardsRemaining).toEqual(5)
		})
	})
})

describe("#killPlayers", () => {
	it("should remove all cards from the targeted players", () => {
		const gameResult = Game.new(["Alice", "Bob", "Charlie", "David"])

		assert(
			Either.isRight(gameResult),
			"Expected Either.right, got Either.left",
		)

		const { right: game } = gameResult

		const updatedGame = game.killPlayers(1, 3)

		const { playersStillAlive } = updatedGame

		expect(playersStillAlive).toHaveLength(2)

		playersStillAlive.forEach((player) => {
			expect(player.name).not.toEqual("Bob")
			expect(player.name).not.toEqual("David")
		})
	})
})

File: core/tests/game_test.rs
use salem_core::{Game, NewGameError};

#[test]
fn should_return_an_error_if_there_are_too_few_players() {
    let error = Game::new(1).expect_err("Expected too few players error");

    assert_eq!(
        error,
        NewGameError::TooFewPlayersError(1),
        "Expected too few players error"
    );
}

#[test]
fn should_give_five_cards_to_each_player_if_there_are_four_or_more_players() {
    let game = Game::new(4).expect("Expected game to be created");
    let players = game.players_still_alive();

    assert_eq!(players.len(), 4, "Expected 4 players");

    for player in players {
        assert_eq!(
            player.tryal_cards_remaining(),
            5,
            "Expected 5 cards per player"
        );
    }
}

#[test]
fn should_give_five_cards_to_each_player_if_there_are_six_or_less_players() {
    let game = Game::new(6).expect("Expected game to be created");
    let players = game.players_still_alive();

    assert_eq!(players.len(), 6, "Expected 6 players");

    for player in players {
        assert_eq!(
            player.tryal_cards_remaining(),
            5,
            "Expected 5 cards per player"
        );
    }
}

#[test]
fn should_give_four_cards_to_each_player_if_there_are_seven_or_more_players() {
    let game = Game::new(7).expect("Expected game to be created");
    let players = game.players_still_alive();

    assert_eq!(players.len(), 7, "Expected 7 players");

    for player in players {
        assert_eq!(
            player.tryal_cards_remaining(),
            4,
            "Expected 4 cards per player"
        );
    }
}

#[test]
fn should_give_four_cards_to_each_player_if_there_are_nine_or_less_players() {
    let game = Game::new(9).expect("Expected game to be created");
    let players = game.players_still_alive();

    assert_eq!(players.len(), 9, "Expected 9 players");

    for player in players {
        assert_eq!(
            player.tryal_cards_remaining(),
            4,
            "Expected 4 cards per player"
        );
    }
}

#[test]
fn should_give_three_cards_to_each_player_if_there_are_ten_or_more_players() {
    let game = Game::new(10).expect("Expected game to be created");
    let players = game.players_still_alive();

    assert_eq!(players.len(), 10, "Expected 10 players");

    for player in players {
        assert_eq!(
            player.tryal_cards_remaining(),
            3,
            "Expected 3 cards per player"
        );
    }
}

#[test]
fn should_give_three_cards_to_each_player_if_there_are_twelve_or_less_players() {
    let game = Game::new(12).expect("Expected game to be created");
    let players = game.players_still_alive();

    assert_eq!(players.len(), 12, "Expected 12 players");

    for player in players {
        assert_eq!(
            player.tryal_cards_remaining(),
            3,
            "Expected 3 cards per player"
        );
    }
}

#[test]
fn should_return_an_error_if_there_are_too_many_players() {
    let error = Game::new(13).expect_err("Expected too many players error");

    assert_eq!(
        error,
        NewGameError::TooManyPlayersError(13),
        "Expected too many players error"
    );
}
